### DISCUSSION TITLE
Cordova error while building android apk

### DISCUSSION THREAD

**Discussion by asgairola-doubtnut** [Sunbird-Ed/Community#7] [Installation] [answered]
npm WARN idealTree Removing dependencies.cordova-plugin-inappbrowser in favor of devDependencies.cordova-plugin-inappbrowser
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: sunbirded-mobile-app@0.0.1
npm ERR! Found: @angular/common@9.1.13
npm ERR! node_modules/@angular/common
npm ERR!   @angular/common@"^9.1.13" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer @angular/common@"^7.1.0" from @project-sunbird/common-consumption-v9@4.4.3
npm ERR! node_modules/@project-sunbird/common-consumption-v9
npm ERR!   @project-sunbird/common-consumption-v9@"4.4.3" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /Users/anurag/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/anurag/.npm/_logs/2022-05-11T08_26_20_326Z-debug-0.log
[ERROR] An error occurred while running subprocess cordova.
        
        cordova plugin add cordova-plugin-inappbrowser exited with exit code 1.
        
        Re-running this command with the --verbose flag may provide more information.
> cordova plugin add cordova-plugin-network-information

**Comment by sharathkashyap** [comment, marked as answer]
@asgairola-doubtnut Can you provide the complete stack trace information. Also kindly provide the node version installed in the system. Please run following commands and provide us with the output.
`ionic info`
`node -v`
There seems to be a conflict in dependencies as per the error posted

**Reply by asgairola-doubtnut** [reply]
Ionic:

   Ionic CLI                     : 5.4.16 (/usr/local/lib/node_modules/ionic)
   Ionic Framework               : not installed
   @angular-devkit/build-angular : not installed
   @angular-devkit/schematics    : not installed
   @angular/cli                  : not installed
   @ionic/angular-toolkit        : not installed

Cordova:

   Cordova CLI       : 11.0.0
   Cordova Platforms : none
   Cordova Plugins   : no whitelisted plugins (0 plugins total)

Utility:

   cordova-res : not installed
   native-run  : not installed

System:

   NodeJS : v16.15.0 (/usr/local/bin/node)
   npm    : 8.5.5
   OS     : macOS Catalina

**Reply by sharathkashyap** [reply]
@asgairola-doubtnut Can you downgrade your node version to 14. Clear your node_modules and do the build once again.

**Reply by asgairola-doubtnut** [reply]
@sharathkashyap after dwongrading to node 14 ... still getting below error 


        Re-running this command with the --verbose flag may provide more information.
> cordova plugin add cordova-plugin-ionic-webview
Failed to fetch plugin cordova-plugin-ionic-webview@^4.2.1 via registry.
Probably this is either a connection problem, or plugin spec is incorrect.
Check your connection and plugin name/version/URL.
CordovaError: Error: npm: Command failed with exit code 1 Error output:
npm WARN idealTree Removing dependencies.cordova-plugin-ionic-webview in favor of devDependencies.cordova-plugin-ionic-webview
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: sunbirded-mobile-app@0.0.1
npm ERR! Found: @angular/common@9.1.13
npm ERR! node_modules/@angular/common
npm ERR!   @angular/common@"^9.1.13" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer @angular/common@"^7.1.0" from @project-sunbird/common-consumption-v9@4.4.3
npm ERR! node_modules/@project-sunbird/common-consumption-v9
npm ERR!   @project-sunbird/common-consumption-v9@"4.4.3" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /Users/anurag/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/anurag/.npm/_logs/2022-05-12T13_22_53_557Z-debug-0.log
[ERROR] An error occurred while running subprocess cordova.
        
        cordova plugin add cordova-plugin-ionic-webview exited with exit code 1.

**Reply by asgairola-doubtnut** [reply]
now node version is v14.19.1

**Reply by asgairola-doubtnut** [reply]
@sharathkashyap also please help us by letting us know the parameters to be added in sunbird.properties 

current file contents are as below:

//App specific Properties
dev_app_id = DEV_APP_ID
staging_app_id = STAGING_APP_ID
production_app_id = PRODUCTION_APP_ID
app_name = APP_NAME
app_version_code = APP_VERSION_CODE
// Dev Enironment properties
dev_base_url = DEV_BASE_URL
dev_mobile_app_key = DEV_MOBILE_APP_KEY
dev_mobile_app_secret = DEV_MOBILE_APP_SECRET
dev_channel_id = DEV_CHANNEL_ID
dev_producer_id = DEV_PRODUCER_ID
dev_deeplink_base_url = DEV_DEEPLINK_BASE_URL
dev_deeplink_ncert_url = DEV_DEEPLINK_NCERT_URL
dev_deeplink_igot_url = DEV_DEEPLINK_IGOT_URL
dev_oauth_redirect_url = DEV_OATH_REDIRECT_URL
// Staging envornment properties.
staging_base_url = STAGING_BASE_URL
staging_mobile_app_key = STAGING_MOBILE_APP_KEY
staging_mobile_app_secret = STAGING_MOBILE_APP_SECRET
staging_channel_id = STAGING_CHANNEL_ID
staging_producer_id = STAGING_PRODUCER_ID
staging_deeplink_base_url = STAGING_DEEPLINK_BASE_URL
staging_deeplink_ncert_url = STAGING_DEEPLINK_NCERT_URL
staging_deeplink_igot_url = STAGING_DEEPLINK_IGOT_URL
staging_oauth_redirect_url = STAGING_OATH_REDIRECT_URL
// Production envornment properties.
production_base_url = PRODUCTION_BASE_URL
production_mobile_app_key = PRODUCTION_MOBILE_APP_KEY
production_mobile_app_secret = PRODUCTION_MOBILE_APP_SECRET
production_channel_id = PRODUCTION_CHANNEL_ID
production_producer_id = PRODUCTION_PRODUCER_ID
production_deeplink_base_url = PRODUCTION_DEEPLINK_BASE_URL
production_deeplink_base_url = PRODUCTION_DEEPLINK_BASE_URL
staging_deeplink_ncert_url = PRODUCTION_DEEPLINK_NCERT_URL
staging_deeplink_igot_url = PRODUCTION_DEEPLINK_IGOT_URL
production_oauth_redirect_url = PRODUCTION_OATH_REDIRECT_URL
// Fabric Keys
release_fabric_api_key = RELEASE_FABRIC_API_KEY
release_fabric_api_secret = RELEASE_FABRIC_API_SECRET
player_deeplink_scheme = PLAYER_DEEPLINK_SCHEME

display_framework_categories_in_profile = DISPLAY_FRAMEWORK_CATEGORIES_IN_PROFILE
display_signin_footer_card_in_course_tab_for_teacher = DISPLAY_SIGNIN_FOOTER_CARD_IN_COURSE_TAB_FOR_TEACHER
display_signin_footer_card_in_library_tab_for_teacher = DISPLAY_SIGNIN_FOOTER_CARD_IN_LIBRARY_TAB_FOR_TEACHER
display_signin_footer_card_in_profile_tab_for_teacher = DISPLAY_SIGNIN_FOOTER_CARD_IN_PROFILE_TAB_FOR_TEACHER
display_signin_footer_card_in_library_tab_for_student = DISPLAY_SIGNIN_FOOTER_CARD_IN_LIBRARY_TAB_FOR_STUDENT
display_signin_footer_card_in_profile_tab_for_student = DISPLAY_SIGNIN_FOOTER_CARD_IN_PROFILE_TAB_FOR_STUDENT
track_user_telemetry = TRACK_USER_TELEMETRY
content_streaming_enabled = CONTENT_STREAMING_ENABLED
open_rapdiscovery_enabled = OPEN_RAPDISCOVERY_ENABLED
display_onboarding_category_page = DISPLAY_ONBOARDING_CATEGORY_PAGE
support_email = SUPPORT_EMAIL

dev_custom_scheme = org.sunbird.app.dev
staging_custom_scheme = staging.diksha.app
production_custom_scheme = in.gov.diksha.app

**Reply by sharathkashyap** [reply]
@asgairola-doubtnut Kindly send an email to us. We will respond back with starter kit which contains Google-services.json and properties file. This would be shared over email to you. Which will provide all the values of these properties. 
Cheers

**Reply by sharathkashyap** [reply]
@asgairola-doubtnut Can you please post the entire stack trace of the build. It seem to have been failing to fetch the plugin. Would need complete trace to identify the issue.

**Reply by asgairola-doubtnut** [reply]
anurag@apples-MacBook-Pro SunbirdEd-mobile-app % ./build.sh
rm: package-lock.json: No such file or directory
✔ Creating ./www directory for you - done!
> cordova plugin add cordova-plugin-printer
Failed to fetch plugin cordova-plugin-printer@^0.8.0 via registry.
Probably this is either a connection problem, or plugin spec is incorrect.
Check your connection and plugin name/version/URL.
CordovaError: Error: npm: Command failed with exit code 1 Error output:
npm WARN idealTree Removing dependencies.cordova-plugin-printer in favor of devDependencies.cordova-plugin-printer
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: sunbirded-mobile-app@0.0.1
npm ERR! Found: @angular/common@9.1.13
npm ERR! node_modules/@angular/common
npm ERR!   @angular/common@"^9.1.13" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer @angular/common@"^7.1.0" from @project-sunbird/common-consumption-v9@4.4.3
npm ERR! node_modules/@project-sunbird/common-consumption-v9
npm ERR!   @project-sunbird/common-consumption-v9@"4.4.3" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /Users/anurag/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/anurag/.npm/_logs/2022-05-16T05_26_20_112Z-debug-0.log

  ────────────────────────────────────────────────────────────

          Ionic CLI update available: 5.4.16 → 6.19.1
                                
     The package name has changed from ionic to @ionic/cli!
                                
             To update, run: npm uninstall -g ionic
                 Then run: npm i -g @ionic/cli

  ────────────────────────────────────────────────────────────

[ERROR] An error occurred while running subprocess cordova.
        
        cordova plugin add cordova-plugin-printer exited with exit code 1.
        
        Re-running this command with the --verbose flag may provide more information.
> cordova plugin add https://github.com/Sunbird-Ed/jjdltc-cordova-plugin-zip.git
Failed to fetch plugin https://github.com/Sunbird-Ed/jjdltc-cordova-plugin-zip.git via registry.
Probably this is either a connection problem, or plugin spec is incorrect.
Check your connection and plugin name/version/URL.
CordovaError: Error: npm: Command failed with exit code 1 Error output:
npm WARN idealTree Removing dependencies.cordova-zip-plugin in favor of devDependencies.cordova-zip-plugin
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: sunbirded-mobile-app@0.0.1
npm ERR! Found: @angular/common@9.1.13
npm ERR! node_modules/@angular/common
npm ERR!   @angular/common@"^9.1.13" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer @angular/common@"^7.1.0" from @project-sunbird/common-consumption-v9@4.4.3
npm ERR! node_modules/@project-sunbird/common-consumption-v9
npm ERR!   @project-sunbird/common-consumption-v9@"4.4.3" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /Users/anurag/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/anurag/.npm/_logs/2022-05-16T05_27_27_797Z-debug-0.log
[ERROR] An error occurred while running subprocess cordova.

@sharathkashyap this is the complete stach trace after I run /build.sh script

**Reply by sharathkashyap** [reply]
@asgairola-doubtnut The issue is none of the plugins are installed from cordova. You may have to try these steps.
Uninstall and Reinstall Cordova Globally on your machine. This step should be done under recommended node version (i.e 12 or 14).
`npm uninstall cordova -g
 npm install cordova -g`
Try using recommended version of cordova. Your version seems to be 11.
Clear the platfforms and do a rebuild.

**Reply by asgairola-doubtnut** [reply]
[ERROR] An error occurred while running subprocess cordova.
        
        cordova platform add android@9.0.0 exited with exit code 1.
        
        Re-running this command with the --verbose flag may provide more information.
  ionic:lib Terminal info: { ci: false, shell: '/bin/zsh', tty: true, windows: false } +0ms
  ionic:lib CLI global options: { _: [ 'cordova', 'build', 'android' ], help: null, h: null, verbose: true, quiet: null, interactive: true, color: true, confirm: null, json: null, prod: true, release: true, buildConfig: './buildConfig/build.json', project: null, '--': [] } +2ms
  ionic:lib:project Project type from config: @ionic/angular (angular) +0ms
  ionic:lib:project Project details: { configPath: '/Users/anurag/SunbirdEd-mobile-app/ionic.config.json', errors: [], context: 'app', type: 'angular' } +0ms
  ionic Context: { binPath: '/usr/local/lib/node_modules/ionic/bin/ionic', libPath: '/usr/local/lib/node_modules/ionic', execPath: '/Users/anurag/SunbirdEd-mobile-app', version: '5.4.16' } +0ms
  ionic:lib:integrations:cordova:config Loading Cordova Config (config.xml: '/Users/anurag/SunbirdEd-mobile-app/config.xml', package.json: '/Users/anurag/SunbirdEd-mobile-app/package.json') +0ms
> cordova platform add android --save
Using cordova-fetch for cordova-android@^9.0.0
Failed to fetch platform cordova-android@^9.0.0
Probably this is either a connection problem, or platform spec is incorrect.
Check your connection and platform name/version/URL.
CordovaError: Error: npm: Command failed with exit code 1 Error output:
2022-05-16T09:49:05.645Z agentkeepalive sock[0#registry.npmjs.org:443::::::::true:::::::::::::] create, timeout 300001ms
2022-05-16T09:49:05.721Z agentkeepalive sock[0#registry.npmjs.org:443::::::::true:::::::::::::](requests: 1, finished: 1) free
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: sunbirded-mobile-app@0.0.1
npm ERR! Found: @angular/common@9.1.13
npm ERR! node_modules/@angular/common
npm ERR!   @angular/common@"^9.1.13" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer @angular/common@"^7.1.0" from @project-sunbird/common-consumption-v9@4.4.3
npm ERR! node_modules/@project-sunbird/common-consumption-v9
npm ERR!   @project-sunbird/common-consumption-v9@"4.4.3" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /Users/anurag/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/anurag/.npm/_logs/2022-05-16T09_49_05_268Z-debug-0.log
[ERROR] An error occurred while running subprocess cordova.
        
        cordova platform add android --save exited with exit code 1.
        
        Re-running this command with the --verbose flag may provide more information.
  ionic:utils-process onBeforeExit handler: 'process.exit' received +0ms
  ionic:utils-process onBeforeExit handler: running 1 functions +0ms
  ionic:utils-process processExit: exiting (exit code: 1) +18ms

> sunbirded-mobile-app@0.0.1 ionic-build
> ng build --eval--source-map --configuration=$npm_config_angular_configuration && cordova build android

sh: ng: command not found
anurag@apples-MacBook-Pro SunbirdEd-mobile-app % 

@sharathkashyap after resintalling cordova again.. i am getting the above error.

**Reply by sharathkashyap** [reply]
@asgairola-doubtnut 
Can you try the following:

Please be on right nvm version before these steps.

Delete node_modules folder - This will remove all dependencies
Delete platforms folder - This should clear out all the possible native folder structure
Delete plugins folder - This should clear all plugin dependecies
Delete package-lock.json file - This will clear out any npm dependencies caxhed
If you use a global cordova installation, then do: npm uninstall -g cordova && npm install -g cordova@9

run npm install
run cordova platform add android

Try these steps in the same order. The issue seems to be around cordova installation.

**Reply by asgairola-doubtnut** [reply]
anurag@apples-MacBook-Pro SunbirdEd-mobile-app % npm install 
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: sunbirded-mobile-app@0.0.1
npm ERR! Found: @angular/common@9.1.13
npm ERR! node_modules/@angular/common
npm ERR!   @angular/common@"^9.1.13" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer @angular/common@"^7.1.0" from @project-sunbird/common-consumption-v9@4.4.3
npm ERR! node_modules/@project-sunbird/common-consumption-v9
npm ERR!   @project-sunbird/common-consumption-v9@"4.4.3" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /Users/anurag/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/anurag/.npm/_logs/2022-05-17T04_43_14_495Z-debug-0.log
anurag@apples-MacBook-Pro SunbirdEd-mobile-app % 


@sharathkashyap while doing npm installl , now this is the stack trace

**Reply by asgairola-doubtnut** [reply]
thanks @sharathkashyap for helping in the detailed steps for installing the same

we were missing below items as discussed 👍 
1. not using latest stable release branch 
2. the npm install in build.sh was not executing
3. placement of the google-services.json at root level
4. the angular installation was not done. 

we are tyring to build the same and will update if things works for us !

**Reply by asgairola-doubtnut** [reply]
@sharathkashyap the apk generated is not installing on our devices... anything that can lead to this ?

**Reply by sharathkashyap** [reply]
@asgairola-doubtnut Glad to Know that you were able to complete build and generate apk. 
1) There is one issue where apk wouldn't get installed if you have other whitelabels of SunbirdEd-Mobile App already installed.Try Uninstalling if you have any.
2) If Point 1) is not the case, need to check through the build logs to see if there was any error thrown during build process.

cc: @luckynvdu

**Reply by asgairola-doubtnut** [reply]
thanks @sharathkashyap after uninstalling the dikhsa app, it worked.

**Reply by sharathkashyap** [reply]
@asgairola-doubtnut I will be locking the activity on this thread. Since all the queries has been answered.

**Comment by kameshbhr** [comment]
@sharathkashyap Why does a separate email required for this? Aren't these publicly accessible?

**Reply by sharathkashyap** [reply]
These files cannot be publicly accessible as it has some configuration specific to environment and app. We are exploring ways on how to make this information available for all contributors.
